Dynatrace/dynatrace-operator
# Add enrichment-rules to client and dynakube

## Description

[K8S-9940](https://dt-rnd.atlassian.net/browse/K8S-9940)

Using the settings API it stores the metadata-enrichment rules in the `DynaKube` when `metadata-enrichment` is enabled.

- It only checks for rules only every `ApiRequestThreshold`
- It uses conditions to store when the last request was done
- If the `metadata-enrichment` feature gets turned off, the status is cleared

Sidenote:
- Tried to restructure the Settings API related logic in a less messy way, for (hopefully) better readability.

## How can this be tested?
0. On your tenant create rules, you need a feature-flag for it to be available, see details in: [K8S-9128 ](https://dt-rnd.atlassian.net/browse/K8S-9128?focusedCommentId=3336492)
   - You need to create the rules for a Kubernetes-cluster monitored-entity, rules defined at the Environment level will no show up (as of right now)
   - You should have a page like this: (follow the "breadcrumbs" in the top left)
![image](https://github.com/Dynatrace/dynatrace-operator/assets/31651557/0c70e180-dcd2-44e0-bd16-0665c34fd6fc)

2. Deploy a `DynaKube` with `metadataEnrichment.enabled: true`
3. Check the status for the rules:
```
apiVersion: dynatrace.com/v1beta2
kind: DynaKube
metadata: ...
spec:
  metadataEnrichment:
    enabled: true
status:
...
  conditions:
  - lastTransitionTime: "2024-06-17T13:38:41Z"
    message: Metadata-enrichment rules are up-to-date in the status
    reason: StatusUpdated
    status: "True"
    type: MetadataEnrichmentRules
  metadataEnrichment:
    rules:
    - enabled: true
      key: cost-center
      mapping: dt.cost.costcenter
      type: LABEL
    - enabled: true
      key: security-context
      mapping: dt.security_context
      type: ANNOTATION
```
